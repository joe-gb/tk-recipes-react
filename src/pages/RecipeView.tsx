import RecipeService, { Ingredient, Recipe } from "../services/Recipe";
import { Link, useParams } from "react-router-dom";
import React from "react";
import { Card } from "../components/Card";
import { SmallSubtitle, Subtitle } from "../components/Text";
import { List, ListItem } from "../components/List";
import { Row } from "../components/Layout";
import { Button } from "../components/Button";
import Loader from "../components/Loader";

function RecipeView() {
  let { id } = useParams();

  let [recipe, loading, error] = RecipeService.GetRecipe(id);

  if (error) {
    return <p>Unable to fetch recipe - {error.message}</p>;
  }

  if (loading) {
    return <Loader />;
  }

  return (
    <Card>
      <Subtitle>{recipe.name}</Subtitle>
      {recipe.ingredients && (
        <>
          <SmallSubtitle>Ingredients</SmallSubtitle>
          {List(
            recipe.ingredients.map((ingredient: Ingredient) => ingredient.name)
          )}
        </>
      )}
      <Row>
        <Link to={"/recipes/edit/" + id}>
          <Button>Edit recipe</Button>
        </Link>
        <Link to={"/recipes/delete/" + id}>
          <Button>Delete recipe</Button>
        </Link>
      </Row>
    </Card>
  );
}

export default RecipeView;
