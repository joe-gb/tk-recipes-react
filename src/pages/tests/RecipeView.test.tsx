import React from "react";
import axios from "axios";
import { BrowserRouter as Router, Routes, Route } from "react-router-dom";
import MockAdapter from "axios-mock-adapter";
import "@testing-library/jest-dom/extend-expect";
import { cleanup, waitFor, render, screen } from "@testing-library/react";
import RecipeView from "../RecipeView";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    id: 5,
  }),
  useRouteMatch: () => ({ url: "/recipes/5" }),
}));

const mockAxios = new MockAdapter(axios);
mockAxios.onGet(/http:\/\/localhost:8000\/recipes\/5\/\?[0-9]+/).reply(200, {
  id: 5,
  name: "Paella",
  ingredients: [
    { name: "Rice" },
    { name: "Prawns" },
    { name: "Paprika" },
    { name: "Tomatoes" },
  ],
});

afterEach(cleanup);

test("Test recipe list", async () => {
  const { container } = render(
    <Router>
      <RecipeView />
    </Router>
  );

  await waitFor(() => {
    expect(screen.getByText("Prawns")).toBeInTheDocument();
  });
  expect(screen.getByText("Rice")).toBeInTheDocument();
  expect(screen.getByText("Paprika")).toBeInTheDocument();
  expect(screen.getByText("Tomatoes")).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
