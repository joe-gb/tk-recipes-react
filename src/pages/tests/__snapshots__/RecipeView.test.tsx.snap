// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test recipe list 1`] = `
.c0 {
  font-size: 32px;
  font-weight: 500;
  -webkit-letter-spacing: -0.5px;
  -moz-letter-spacing: -0.5px;
  -ms-letter-spacing: -0.5px;
  -webkit-letter-spacing: -0.5px;
  -moz-letter-spacing: -0.5px;
  -ms-letter-spacing: -0.5px;
  letter-spacing: -0.5px;
  line-height: 40px;
  text-align: left;
  vertical-align: baseline;
  text-transform: unset;
  color: #1d2c3c;
  margin: 20px auto;
  max-width: 850px;
}

.c1 {
  font-size: 26px;
  font-weight: 500;
  -webkit-letter-spacing: -0.5px;
  -moz-letter-spacing: -0.5px;
  -ms-letter-spacing: -0.5px;
  -webkit-letter-spacing: -0.5px;
  -moz-letter-spacing: -0.5px;
  -ms-letter-spacing: -0.5px;
  letter-spacing: -0.5px;
  line-height: 40px;
  text-align: left;
  vertical-align: baseline;
  text-transform: unset;
  color: #1d2c3c;
  margin: 20px auto;
  max-width: 850px;
}

.c2 {
  font-size: 18px;
  font-weight: 400;
  line-height: 24px;
  text-align: left;
  vertical-align: baseline;
  text-transform: unset;
  color: #4c5566;
  background-color: #eff2f7;
  margin: 2px 0px 2px 0px;
  padding: 10px 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  gap: 10px;
}

.c3 * {
  width: 100%;
}

.c4.c4 {
  display: block;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: 100%;
  margin: 20px auto;
  cursor: pointer;
  font-size: 16px;
  font-weight: 500;
  line-height: 18px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #fff;
  border-radius: 8px;
  background: #003097;
  border: 1px solid #003097;
  padding: 19px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  -webkit-transition: all 0.175s ease-in-out;
  transition: all 0.175s ease-in-out;
}

.c4.c4:hover {
  background-color: #144fcc;
  border: 1px solid #144fcc;
}

<div>
  <h2
    class="c0"
  >
    Paella
  </h2>
  <h3
    class="c1"
  >
    Ingredients
  </h3>
  <div
    class="c2"
  >
    Rice
  </div>
  <div
    class="c2"
  >
    Prawns
  </div>
  <div
    class="c2"
  >
    Paprika
  </div>
  <div
    class="c2"
  >
    Tomatoes
  </div>
  <div
    class="c3"
  >
    <a
      href="/recipes/edit/5"
    >
      <button
        class="c4"
      >
        Edit recipe
      </button>
    </a>
    <a
      href="/recipes/delete/5"
    >
      <button
        class="c4"
      >
        Delete recipe
      </button>
    </a>
  </div>
</div>
`;
